src/transport/raw/TCP.cpp

```diff
@@ -206,53 +201,39 @@ CHIP_ERROR TCPBase::SendMessage(const Transport::PeerAddress & address, System::
 CHIP_ERROR TCPBase::SendAfterConnect(const PeerAddress & addr, System::PacketBufferHandle && msg)
 {
     // This will initiate a connection to the specified peer
-    CHIP_ERROR err               = CHIP_NO_ERROR;
-    PendingPacket * packet       = nullptr;
     bool alreadyConnecting       = false;
-    Inet::TCPEndPoint * endPoint = nullptr;
 
     // Iterate through the ENTIRE array. If a pending packet for
     // the address already exists, this means a connection is pending and
     // does NOT need to be re-established.
-    for (size_t i = 0; i < mPendingPacketsSize; i++)
-    {
-        if (mPendingPackets[i].packetBuffer.IsNull())
-        {
-            if (packet == nullptr)
-            {
-                // found a slot to store the packet into
-                packet = mPendingPackets + i;
-            }
-        }
-        else if (mPendingPackets[i].peerAddress == addr)
+    mPendingPackets.ForEachActiveObject([&] (PendingPacket * pending) {
+        if (pending->mPeerAddress == addr)
         {
             // same destination exists.
             alreadyConnecting = true;
-
-            // ensure packets are ORDERED
-            if (packet != nullptr)
-            {
-                packet->peerAddress  = addr;
-                packet->packetBuffer = std::move(mPendingPackets[i].packetBuffer);
-                packet               = mPendingPackets + i;
-            }
+            pending->mPacketBuffer->AddToEnd(std::move(msg));
```

Comment:
@kpschoedel Here, instead of occupying another slot in the pending queue, I just append the packet to the end of the original packet buffer. I have checked the implementation of inet/`TCPEndpoint::Send`, it seems it should work this way.

this should only work if the packet already includes length information, I'm not sure where and how is tcp stream cut int packets. so I'm not totally confident if this is legit.

Reply:
I think this is OK (but I don't know the depths of the LwIP version.)

Reply:
Checked that tcp data is cut at transport layer `TCPBase::ProcessReceivedBuffer`, It is fine to concat them in inet layer.